blaze-core/src/main/scala/org/http4s/blazecore/websocket/Http4sWSStage.scala

```diff
@@ -152,25 +153,26 @@ private[http4s] class Http4sWSStage[F[_]](
         .compile
         .drain
 
-    unsafeRunAsync(wsStream) {
+    val result = F.attempt(wsStream).flatMap {
```

Comment:
Current situation: `F.attempt(wsStream)` does not make any progress.
In the previous implementation (`unsafeRunAsync(wsStream)`), `wsStream` was "shift" to the execution context provided in the constructor.

Reply:
I'm very far out of the CE3 loop right now.  Do we need a `start` or something similar?

Reply:
The `Dispatcher` can start it. See at line 165: `D.unsafeRunSync(result)`.
When I add a timeout at this line, the timeout exception is coming from this place.

Reply:
`unsafeRunSync` is blocking, so the semantics have changed here. Try `unsafeToFuture` instead of `unsafeRunSync`, or add `.start` at the end of `result`

Reply:
also the `attempt` can become an `handleErrorWith`

Reply:
The previous implementation is this one:
```
  private[http4s] def unsafeRunAsync[F[_], A](fa: F[A])(
      f: Either[Throwable, A] => IO[Unit])(implicit F: Effect[F], ec: ExecutionContext): Unit =
    F.runAsync(Async.shift(ec) *> fa)(f).unsafeRunSync()
```
Does it wait for the result?
Or does it "just" start the computation?

Reply:
`runAsync` doesn't wait, that's why you need to pass the `f` callback, instead of just receiving an `Either[Throwable, A]` as a result (and that's why it's called `*async`), but it doesn't `.start` either (the IO will shift itself as soon as an async node is encountered, which is why we add an explicit `shift` node just before). 

The closest thing to that is `unsafeToFuture`.

Reply:
OK I understand about the `F.runAsync`.
I still have questions about the `.unsafeRunSync()`, if this waits for the start.
If yes `unsafeRunAsync` starts the side-effect on another thread pool right?
As `unsafeToFuture` returns a `Future`, should we wait for the `Future` with an `Await.result` or something similar? In that case, we need a timeout that we don't have in `unsafeRunAsync` .

But if `.unsafeRunSync()` does not wait for the start, then we can start the `Future` and ignore it, just returning Unit.

Reply:
> I still have questions about the .unsafeRunSync(), if this waits for the start.

it waits for the completion of the portion of IO preceding the first async boundary. In this case, because there is a `shift` immediately, it essentially doesn't wait. (and this is in CE2 anyway)

> If yes unsafeRunAsync starts the side-effect on another thread pool right?

Strictly speaking, it just follows what the IO does, and will follow whatever async boundary the IO has. In this case there is an immediate shift.

> As unsafeToFuture returns a Future, should we wait for the Future with an Await.result or something similar? 

no, we don't need to `Await.result`, and in fact what you're doing right now with `unsafeRunSync` is essentially `Await.result`,  which I think it's causing the deadlock in the test.

>  then we can start the Future and ignore it, just returning Unit.

yes

Reply:
In fact, call `unsafeRunAndForget` on `Dispatcher`

Reply:
@SystemFw thank you a lot for your detailed answers ❤️

Reply:
Thank you for porting this :)

Reply:
https://github.com/http4s/http4s/pull/3894/commits/ed7c215810da311fdbaa8b1b911ef717fcb0cf7b
https://github.com/http4s/http4s/pull/3894/commits/932bd3a761fd173bba81f77c81ab756e24b35a4a

so that the tests could run, I had to also do the following: https://github.com/http4s/http4s/pull/3894/commits/f7d5840e441deb56c22e1bcaf7f8e382107609f4